File: internal/dsl/token/token.go
package token

type TokenType string

type Token struct {
	Type    TokenType
	Literal string
	Line    int
	Col     int
}

const (
	EOF     TokenType = "EOF"
	ILLEGAL TokenType = "ILLEGAL"

	IDENT    TokenType = "IDENT"
	BACKTICK TokenType = "BACKTICK"
	PATH_LIT TokenType = "PATH"

	ASSIGN    TokenType = "="
	LBRACE    TokenType = "{"
	RBRACE    TokenType = "}"
	LBRACKET  TokenType = "["
	RBRACKET  TokenType = "]"
	LPAREN    TokenType = "("
	RPAREN    TokenType = ")"
	COMMA     TokenType = ","
	DOT       TokenType = "."
	SEMICOLON TokenType = ";"
	DOLLAR    TokenType = "$"

	SANCTUARY TokenType = "SANCTUARY"
	IMPORT    TokenType = "IMPORT"
	VAR       TokenType = "VAR"
	STRING_KW TokenType = "STRING_KW"
	PR        TokenType = "PR"
	FN        TokenType = "FN"
	SEQ       TokenType = "SEQ"
	PAR       TokenType = "PAR"
	ENV       TokenType = "ENV"
	LOG       TokenType = "LOG"
	EXEC      TokenType = "EXEC"
	CD        TokenType = "CD"
	SHELL     TokenType = "SHELL"
)

var keywords = map[string]TokenType{
	"sanctuary": SANCTUARY,
	"import":    IMPORT,
	"var":       VAR,
	"string":    STRING_KW,
	"pr":        PR,
	"fn":        FN,
	"seq":       SEQ,
	"par":       PAR,
	"env":       ENV,
	"log":       LOG,
	"exec":      EXEC,
	"cd":        CD,
	"shell":     SHELL,
}

func LookupIdent(ident string) TokenType {
	if t, ok := keywords[ident]; ok {
		return t
	}
	return IDENT
}

File: src/dsl/token.rs
#[derive(Debug, Clone, PartialEq)]
#[allow(clippy::upper_case_acronyms)]
pub enum TokenType {
    EOF,
    Illegal(String),
    Ident(String),
    Backtick(String),
    LBrace,
    RBrace,
    LParen,
    RParen,
    LBracket,
    RBracket,
    Semicolon,
    Comma,
    Dot,
    Assign,
    Dollar,
    Import,
    Shell,
    Sanctuary,
    Var,
    StringKw,
    Fn,
    Seq,
    Par,
    Pr,
    Log,
    Exec,
    Cd,
    Env,
    PathLit(String),
}

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub ty: TokenType,
    pub line: usize,
    pub col: usize,
    pub offset: usize,
    pub len: usize,
}

impl Token {
    pub fn new(ty: TokenType, line: usize, col: usize, offset: usize, len: usize) -> Self {
        Self {
            ty,
            line,
            col,
            offset,
            len,
        }
    }
}

pub fn lookup_ident(ident: &str) -> TokenType {
    match ident {
        "sanctuary" => TokenType::Sanctuary,
        "import" => TokenType::Import,
        "var" => TokenType::Var,
        "string" => TokenType::StringKw,
        "pr" => TokenType::Pr,
        "fn" => TokenType::Fn,
        "seq" => TokenType::Seq,
        "par" => TokenType::Par,
        "env" => TokenType::Env,
        "log" => TokenType::Log,
        "exec" => TokenType::Exec,
        "cd" => TokenType::Cd,
        "shell" => TokenType::Shell,
        _ => TokenType::Ident(ident.to_string()),
    }
}
